src/array/domain.rs

```diff
@@ -27,14 +28,31 @@ impl Drop for RawDomain {
 pub struct Domain<'ctx> {
     context: &'ctx Context,
     raw: RawDomain,
-    dimensions: Vec<Dimension<'ctx>>,
+
+    /// Array dimensions attached to this domain.
+    /// These must not be freed until the Domain is.
+    /// UnsafeCell is used for interior mutability so that when a dimension
+    /// is queried we can stash it here and
+    dimensions: UnsafeCell<Vec<Option<Dimension<'ctx>>>>,
```

Comment:
First, the lifetime description is wrong. The pointer you get from core when you call `tiledb_domain_get_dimension_from_index` is to a wrapper handle. It can be deallocated at any time before the owning context is destroyed. The actual internal object is behind a shared_ptr on the pointer we deallocate in Drop traits.

Secondly, I'd probably not attempt to store the dimensions here as a cache. Just have `domain.dimension(index) -> TileDBResult<Dimension<'ctx>>` or w/e.

Reply:
Coming back to re-stress that we should *really* avoid keeping a second state like this. I've never seen it lead anywhere except to subtle bugs when the wrapped copy of state drifts from what the interior state of the domain instance contains.

In these cases, each time we get a dimension, all we're actually getting is a handle class that contains a shared_ptr to the "real" dimension internal to the domain. And those handles are designed to be created cheaply and destroyed frequently *specifically* to reduce the refcounting/lifetime issues in language bindings.

Reply:
I'm in agreement that the less state we keep the better.

I'm going to double check this but the reason that I put the member here in the first place is that I was getting memory errors in the quickstart_dense example with array creation when I first got the API far along enough.  I'm pretty sure approaching it this way fixed it.  I'll check that now because obviously it matters a lot for what the path forward here is.

Reply:
All righty, so... git bisect points at
`58dd0e6dc1ae6b347f050825888d6d75f5083fda is the first bad commit`

where "bad" actually means "does not segfault".

This is the commit where I put the `Vec<Dimension>` in there and so on.

If we look at the C API of `quickstart_dense.c` then we see that the `create_array` function allocates objects, processes them, and then frees them all at the end.  The "good", i.e. faulting, commits do not follow this pattern because they drop the Rust API structs eagerly, e.g.

```
let domain = {
    /* dimensions are created and dropped, and thus freed, within this block, resulting in SEGV */
};
```
whereas moving the dimensions into the `Vec<Dimension>` field of Domain did not drop them until after the domain is done being used.